File: src/cdisc_library_api/specs/ct_codelist.clj
(ns cdisc-library-api.specs.ct-codelist
  (:require [clojure.spec.alpha :as s]
            [spec-tools.data-spec :as ds]
            [cdisc-library-api.specs.ct-codelist-links :refer :all]
            [cdisc-library-api.specs.ct-package-term :refer :all]
            )
  (:import (java.io File)))


(def ct-codelist-data
  {
   (ds/opt :conceptId) string?
   (ds/opt :extensible) string?
   (ds/opt :name) string?
   (ds/opt :submissionValue) string?
   (ds/opt :definition) string?
   (ds/opt :preferredTerm) string?
   (ds/opt :synonyms) (s/coll-of string?)
   (ds/opt :_links) ct-codelist-links-spec
   (ds/opt :terms) (s/coll-of ct-package-term-spec)
   })

(def ct-codelist-spec
  (ds/spec
    {:name ::ct-codelist
     :spec ct-codelist-data}))

File: src/main/scala/org/openapitools/client/api/CtCodelist.scala
package org.openapitools.client.api

import argonaut._
import argonaut.EncodeJson._
import argonaut.DecodeJson._

import org.http4s.{EntityDecoder, EntityEncoder}
import org.http4s.argonaut._
import org.joda.time.DateTime


import CtCodelist._

case class CtCodelist (
  conceptId: Option[String],
extensible: Option[String],
name: Option[String],
submissionValue: Option[String],
definition: Option[String],
preferredTerm: Option[String],
synonyms: Option[List[String]],
links: Option[CtCodelistLinks],
terms: Option[List[CtPackageTerm]])

object CtCodelist {
  import DateTimeCodecs._

  implicit val CtCodelistCodecJson: CodecJson[CtCodelist] = CodecJson.derive[CtCodelist]
  implicit val CtCodelistDecoder: EntityDecoder[CtCodelist] = jsonOf[CtCodelist]
  implicit val CtCodelistEncoder: EntityEncoder[CtCodelist] = jsonEncoderOf[CtCodelist]
}
